### PR TITLE
fix(stats): unsubscribe from stats subscription when leaving dashboard

### PR DESCRIPTION
**Pull request by Izak88**

### CODE REVIEW

File: src/api/docker-stats.ts [hunk 1]
```diff
@@ -52,6 +51,10 @@ export function getContainersStats(): Observable<any> {
       })
       .map(stats => observer.next({ type: 'containersStats', data: stats }))
       .subscribe();
+
+    return () => {
+      sub.unsubscribe();
```
Comment (jkuri) [inline src/api/docker-stats.ts]: please check if `sub` exists before unsubscribing, just in case that we prevent errors.

**Comment by Izak88** [comment]
updated, please for a new review...

File: src/api/docker-stats.ts [hunk 2]
```diff
@@ -1,60 +1,69 @@
 import { Observable, Observer } from 'rxjs';
 import * as utils from './utils';
 import * as dockerode from 'dockerode';
```
Comment (jkuri) [inline src/api/docker-stats.ts]: remove `dockerode` from that file. should be included only in `docker.ts`

File: src/api/docker-stats.ts [hunk 3]
```diff
@@ -1,60 +1,69 @@
...
+    return () => {
+      if (sub) {
+        sub.unsubscribe();
+      }
+    };
   }).share();
 }
 
+function calculateContainerStats(container: dockerode.ContainerInfo): Promise<any> {
+  return docker.getContainer(container.Id).stats()
```
Comment (jkuri) [inline src/api/docker-stats.ts]: ...and that function should be in `docker.ts` as well.

File: src/api/socket.ts [hunk 4]
```diff
@@ -41,6 +41,7 @@ export interface Client {
   session: { cookie: any, ip: string, userId: number, email: string, isAdmin: boolean };
   socket: uws.Socket;
   send: Function;
+  subscriptions: { stats: Subscription };
```
Comment (jkuri) [inline src/api/socket.ts]: 👍

File: src/api/docker-stats.ts (hunk 3 above)
Comment (Izak88) [reply src/api/docker-stats.ts]: done

File: src/api/docker-stats.ts (hunk 2 above)
Comment (Izak88) [reply src/api/docker-stats.ts]: done

File: src/api/docker-stats.ts [hunk 5]
```diff
@@ -1,60 +1,46 @@
...
+          id: st.id,
+          name: st.name,
+          cpu: getCpuData(st.data),
+          network: getNetworkData(st.data),
+          memory: getMemory(st.data),
+          debug: st.debug
+        };
+      });
+
+      Promise.resolve(stats);
```
Comment (jkuri) [inline src/api/docker-stats.ts]: you don't need to do `Promise.resolve`, basic `return stats;` or `return stats.map(st => ....);` would be right.

File: src/api/docker.ts [hunk 6]
```diff
@@ -255,3 +259,40 @@ export function isDockerInstalled(): Observable<boolean> {
     });
   });
 }
+
+export function calculateContainerStats(
```
Comment (jkuri) [inline src/api/docker.ts]: fix that style and remember how we handle that:
```ts
export function calculateContainerStats(
  container: dockerode.ContainerInfo,
  processes: any
): Promise<any> {
  return docker.getContainer(container.id).stats()
    .then(stream => { ....

    });
}
```
then please keep using that style in the future.

**Comment by Izak88** [comment]
updated, after CR

**Pull request merged**